services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProvider.java

```diff
@@ -363,23 +365,46 @@ public Response export(UriInfo uriInfo, RealmModel realm, String format) {
             String nameIDPolicyFormat = getConfig().getNameIDPolicyFormat();
 
 
-            List<Element> signingKeys = streamForExport(session.keys().getKeysStream(realm, KeyUse.SIG, Algorithm.RS256), false)
+            List<KeyDescriptorType> signingKeys = session.keys().getKeysStream(realm, KeyUse.SIG, Algorithm.RS256)
+                    .filter(key -> key.getCertificate() != null)
```

Comment:
Is there any reason for not filtering active keys for signing as we do for encrypted keys?

Reply:
Where do we filter active keys for encrypted keys?

Reply:
I see what you mean now. Enabled and active are not the same thing.

We do not filter enabled for signing keys because method `Stream<KeyWrapper> getKeysStream(RealmModel realm, KeyUse use, String algorithm)` does this already, while method `Stream<KeyWrapper> getKeysStream(RealmModel realm)` does not. 

This is something that could be fixed, however, it would require going through all usages and making it consistent which I think should not be part of this PR as it is already quite big.

On the other hand, exporting only active keys could also make sense, but I am not sure what problems  it can cause as at the moment we export both active and passive. @mposolda @stianst Should we export also passive keys as part of SP metadata or there should be only active ones?

Reply:
@mhajas @cgeorgilakis Good question. I think it is correct that export for "sig" returns also passive keys, but for "enc" it returns just active keys.

Few more details:
it works in a way that Keycloak always has 1 active key per algorithm. So for example in OIDC, there can be this scenario related to key rotation:
1) Keycloak signs token with active key `abc` and send to javascript client "foo"
2) Keycloak administrator rotates keys and sets the new key `def` for algorithm `RS256` as new active. The key `abc` is still kept as "passive" key
3) At some point (EG. after 10 seconds), javascript client sends the token from step1 to REST service to access some data
4) The REST service wants to verify token, which was signed in step 1 by key `abc`. It want to download realm keys to be able to verify signature. The key `abc` needs to be still present even if it is not "active" as the token was signed with that before

I guess that for SAML IDP, this is likely not so important as Keycloak signs just SAMLRequest sent to the SAML IDP, (or SAMLResponse for some rare cases like logout) AFAIK? Still if we return only "active" keys, I can imagine something like this:
1) Keycloak sends SAMLRequest to IDP signed by key `abc`
2) In the meantime, key is rotated in Keycloak and active key is `abc`
3) SAML IDP wants to verify signature and hence it downloads the keys from SP metadata endpoint. It won't see key key with `abc` and hence it won't be able to verify signature on SAMLRequest and hence it will throw an error -> KO.

So it is likely more proper approach to return also "passive" keys for signing, so the key `abc` is returned from SP metadata and SAML IDP is able to verify signature.

Regarding encryption/decryption, we should likely return only "active" keys as this is what SAML IDP needs to use for "encrypt" SAML assertion. But during verification, we should use also "passive" keys due the rotation. For example scenario like:
1) Key `abc` is exported in SP metadata
2) Based on that, SAML Identity provider will encrypt the assertion with key `abc`
3) In the meantime, keys are rotated in Keycloak and `abc` becomes "passive" key
4) Keycloak won't be able to decrypt the assertion with key `abc` because it is not active -> KO

Could the last scenario happen? It seems to me that not as SAMLDecryptionKeysLocator introduced in this PR considers also "passive" keys for SAML assertion decryption, which is correct.

So we should be good with the changes in this PR.

Reply:
Thanks for the detailed response @mposolda. Based on your response I believe we should change this PR as currently we include also passive encryption keys in SP metadata and I agree with you that it makes sense to include only active ones. 

I will update the PR.

Reply:
Ok, Thanks for the updates and doublecheck the codebase and for additional comments in the code!